### PR TITLE
Refactor `BaseSensorOperator` introduce `skip_policy` parameter

### PR DESCRIPTION
**Pull request by raphaelauv**
fix: https://github.com/apache/airflow/pull/40915

- remove `never_fail` parameter 
- introduce `skip_policy` parameter 
- deprecate `soft_fail` and `silent_fail` parameters

### CODE REVIEW

**Comment by raphaelauv** [comment]
@Lee-W @shahar1 @eladkal wdyt ? thanks

File: airflow/sensors/base.py
Comment (shahar1) [inline airflow/sensors/base.py]: Comment is a unclear

File: airflow/sensors/base.py
Comment (shahar1) [inline airflow/sensors/base.py]: Description needs to be added, probably with details about each option in the enum.
On the way, I'd include answers to the following questions:
1. What is a "soft" error?
2. What's the difference between "ignore" and "skip"? (it might not be as trivial for the user)

File: airflow/sensors/base.py
Comment (shahar1) [inline airflow/sensors/base.py]: 1. Mutual exclusiveness with `silent_fail` should be clarified.
2. It should be marked as deprecated.

File: airflow/sensors/base.py
Comment (shahar1) [inline airflow/sensors/base.py]: Namings might need some minor improvements, for example:

```suggestion
    SKIP_ON_SOFT_ERROR = "skip_on_soft_error"
```

**Review by shahar1 — COMMENTED** [review]
Looks good overall - had some comments, most of them are regarding documenation and terminology.
@Lee-W - I'd be happy if you could also review it.

File: airflow/sensors/base.py
Comment (eladkal) [inline airflow/sensors/base.py:217]: Im curious what if user has custom skip exception?
Can the user add support for it from deployment side?

**Review by eladkal — APPROVED** [review]
Nice!

**Review by shahar1 — APPROVED** [review]

File: airflow/sensors/base.py
Comment (Lee-W) [inline airflow/sensors/base.py:121]: I feel we might need to update documentation for this as well

File: airflow/sensors/base.py
Comment (Lee-W) [inline airflow/sensors/base.py:136]: I'm wondering whether we could reword this. It looks similar to `SKIP_ON_ANY_ERROR` at the first glance 🤔

File: airflow/sensors/base.py
Comment (Lee-W) [inline airflow/sensors/base.py:132]: I might suggest rewording this one as well. The description here is great, but `soft`_fail sounds a bit unclear to me

File: airflow/sensors/base.py
Comment (Lee-W) [inline airflow/sensors/base.py:218]: If we're already using `skip_policy` to check. Do we still need this assignment here?

File: airflow/sensors/base.py
Comment (raphaelauv) [reply airflow/sensors/base.py:132]: ok , what is your proposition ?

File: airflow/sensors/base.py
Comment (raphaelauv) [reply airflow/sensors/base.py:136]: this is a copy past of the definition of `silent_fail` , could you please make a suggestion , thanks

File: airflow/sensors/base.py
Comment (raphaelauv) [reply airflow/sensors/base.py:121]: yes I will in another PR , when we agree on the implementation , cause I don't want to do big refacto every time we change the terminology

File: airflow/sensors/base.py
Comment (raphaelauv) [reply airflow/sensors/base.py:218]: some sensors still use self.soft_fail

File: airflow/sensors/base.py
Comment (raphaelauv) [reply airflow/sensors/base.py:217]: users already repack custom exceptions https://github.com/apache/airflow/blob/e9d2d5c86a714d6aea7c4dfb3b8cf72e54b9035c/airflow/providers/sftp/sensors/sftp.py#L103

```Can the user add support for it from deployment side```

It feel like this is another feature , we can create an issue to track that

**Comment by raphaelauv** [comment]
@eladkal should we merge this part and continue to iterate with the next PR's or wait for another review of @Lee-W  ? thanks

File: airflow/sensors/base.py
Comment (Lee-W) [reply airflow/sensors/base.py:136]: Hi, sorry for the late reply. Nothing comes to my mind as of now. As long as we document it, I think we're good. Thanks 🙂 Resolve this conversation.

File: airflow/sensors/base.py
Comment (Lee-W) [reply airflow/sensors/base.py:132]: Hi, sorry for the late reply. Nothing comes to my mind as of now. As long as we document it, I think we're good. Thanks 🙂 Resolve this conversation.

**Comment by raphaelauv** [comment]
@eladkal gentle ping ( I would like to finish this work for airflow 2.10 , so we can trust sensors again in airflow , cause right now we can't have at least once pattern with the soft_fail option )

**Review by eladkal — APPROVED** [review]

**Comment by potiuk** [comment]
Hey @raphaelauv -> This one broke our canary builds https://github.com/apache/airflow/actions/runs/10107479883/job/27951619130 -> can you also apply the new policy to all the operators that are raising warnings now ?

**Comment by potiuk** [comment]
Warnings detected in test case(s):
 - tests/providers/apache/flink/sensors/test_flink_kubernetes.py::TestFlinkKubernetesSensor::test_cluster_error_state[True-AirflowSkipException]
 - tests/providers/cncf/kubernetes/sensors/test_spark_kubernetes.py::TestSparkKubernetesSensor::test_driver_logging_failure[True-AirflowSkipException]
 - tests/providers/cncf/kubernetes/sensors/test_spark_kubernetes.py::TestSparkKubernetesSensor::test_failed_application[True-AirflowSkipException]
 - tests/providers/cncf/kubernetes/sensors/test_spark_kubernetes.py::TestSparkKubernetesSensor::test_unknown_application[True-AirflowSkipException]
 - tests/providers/datadog/sensors/test_datadog.py::TestDatadogSensor::test_sensor_fail_with_exception[True-AirflowSkipException]
 - tests/providers/http/sensors/test_http.py::TestHttpSensor::test_poke_exception_with_soft_fail

**Comment by raphaelauv** [comment]
@potiuk yes I started the replace in this PR -> https://github.com/apache/airflow/pull/40923

i'm going to finish in few hours

**Comment by potiuk** [comment]
BTW, This should be made compatible with Airflow 2.7 + (maybe we could use common.compat provider.

In tne meantime, I will revert this one and maybe it will be better if you combine the two together this time ?

File: airflow/sensors/base.py
Comment (raphaelauv) [reply airflow/sensors/base.py:132]: what about SKIP_ON_TIMEOUT and remove AirflowFailException from the list of exception catch by this option

naming would be explicit and correct

wdyt ?

**Pull request merged**
